ihmc-perception/src/test/java/us/ihmc/perception/cuda/examples/ExampleCUDAKernel2D.java

```diff
@@ -0,0 +1,124 @@
+import us.ihmc.perception.cuda.CUDAKernel;
+import us.ihmc.perception.cuda.CUDAProgram;
+import us.ihmc.perception.cuda.CUDAStreamManager;
+
+import java.net.URISyntaxException;
+import java.nio.file.Path;
+import java.util.Objects;
+
+/**
+ * This is an example of using a more complex CUDA kernel that takes in a JavaCV {@link Mat} object. We do this often for kernels, so this example walks you
+ * through how it would work for a simple case.
+ * This uses more of the standard approach then the {@link ExampleCUDAKernel} did because we use the {@link CUDAStreamManager} and we are creating objects
+ * in the try-with-resources statement.
+ */
+public class ExampleCUDAKernel2D
+{
+   public ExampleCUDAKernel2D() throws URISyntaxException
+   {
+      // We load the kernel from resources; this is a good place to store kernels to separate them from the Java classes
+      Path programPath = Path.of(Objects.requireNonNull(getClass().getResource("matrix_element_wise_addition.cu")).toURI());
+      // Create the stream
+      CUstream_st stream = CUDAStreamManager.getStream();
+
+      // Here we want to maximize the number of threads we can, this helps optimize the kernel for runtime.
+      // By the rules of CUDA, a block cannot have more than 1024 threads.
+      // So since we are using a matrix, we are going to have a height and a width, so those squared can't be more the 1024.
+      // Here we expect a grid of 3x3
+      int height = (int) Math.sqrt(9.0);
+      int width = (int) Math.sqrt(9.0);
+
+      // The following commented code below will fail because it will try to create more than 1024 threads within one block, try it...
+      //      int height = (int) Math.sqrt(1024.0) + 1;
+      //      int width = (int) Math.sqrt(1024.0) + 1;
+
+      // Note this name DOES have to match the name of the CUDA kernel you want to run, that means the name of the method.
+      //  It's not guaranteed to be the name of the file either if the ( .cu ) file contains more than one method
+      String kernelName = "element_wise_add";
+
+      // This is where we create all of our objects that will be used in the kernel, the following formatting will be gross because of all the comments
+      try (
+            // This creates the inputs to the kernel, as well as the cpu result we expect to get back and use in Java
+            Mat cpuMatA = new Mat(height, width, opencv_core.CV_16UC1, new Scalar(3));
+            Mat cpuMatB = new Mat(height, width, opencv_core.CV_16UC1, new Scalar(5));
+            Mat cpuResult = new Mat();
+
+            // This creates objects that will be allocated on the gpu, these will be used in the kernel since that all happens on the gpu
+            GpuMat gpuMatA = new GpuMat();
+            GpuMat gpuMatB = new GpuMat();
+            GpuMat gpuResult = new GpuMat(cpuMatA.size(), cpuMatA.type());
+            // We need to know the size in bytes and the type of the result we expect to get on the gpu
```

Comment:
Sorry, I never explained this that well. The `size()` method returns a `Size` object that contains the width and height of the `Mat`. Not super directly related to the size in bytes. `type()` returns the data type (as in, 16 bit unsigned shorts). SO internally the constructor DOES calculate byteSize = width * height * bytesPerPixel

Reply:
Ya thanks, sound good